src/App.tsx

```diff
@@ -1,24 +1,201 @@
+import { TodoList } from './components/TodoList';
+import { FilterOptions } from './types/FilterOptions';
+import { Todo } from './types/Todo';
+import {
+  getTodos, createTodo, removeTodo, updateTodo,
+} from './api/todos';
 import { UserWarning } from './UserWarning';
+import { filterTodos } from './utils/helpers';
+import { Footer } from './components/Footer';
 
-const USER_ID = 0;
+const USER_ID = 10873;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState<FilterOptions>(FilterOptions.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodos, setLoadingTodos] = useState([0]);
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+  const isTodosFormServer = todos.length > 0;
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const todosFromServer: Todo[] = await getTodos(USER_ID);
+
+        setTodos(todosFromServer);
+      } catch {
+        setError('Unable to get todos');
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  useEffect(() => {
+    let timeoutId: number;
+
+    if (error) {
+      timeoutId = window.setTimeout(() => setError(null), 3000);
+    }
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [error]);
+
+  const visibleTodos = useMemo(() => {
+    return filterTodos(filter, todos, activeTodos, completedTodos);
+  }, [todos, filter]);
+
+  const addTodo = async (title: string) => {
+    try {
+      const newTodo = {
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const updateTodoInfo = async (
+    todoId: number,
+    newTodoData: Partial<Pick<Todo, 'title' | 'completed'>>,
+  ) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoId]);
+      const updatedTodo = await updateTodo(todoId, newTodoData);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return updatedTodo;
+      }));
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodos(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  };
+
+  const onUpdateTodo = async (todoId: number, newTitle: string) => {
+    await updateTodo(todoId, { title: newTitle });
+  };
+
+  const onUpdateTodoStatus = async (todo: Todo) => {
+    await updateTodo(todo.id, { completed: !todo.completed });
+  };
```

Comment:
When I click on a todo checkbox I don't see any changes.
I can see them only after the page is reloaded. 

![image](https://github.com/mate-academy/react_todo-app-with-api/assets/84296041/b652ca54-c0d5-43d1-97eb-ebb660ba1a6a)
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/84296041/688aabe6-26cf-4a84-9f1a-5a4e8bd0334f)


I think it should use similar logic to `updateTodoInfo`